File: AMFilter3D.py
# TRANSLATED PYTHON FUNCTION FOR AMFilter.m by LANGELAAR
import numpy as np
import numpy.matlib
import scipy.sparse as sps
from copy import deepcopy

def AMFilter(x, baseplate, *args):
    #   Possible uses:
    #   xi = AMfilter(x, baseplate)   idem, with baseplate orientation specified
    #   [xi, df1dx, df2dx,...] = AMfilter(x, baseplate, df1dxi, df2dxi, ...)
    #       This includes also the transformation of design sensitivities
    # where
    #   x : blueprint design (2D array), 0 <= x(i,j) <= 1
    #   xi: printed design (2D array)
    #   baseplate: character indicating baseplate orientation: 'N','E','S','W'
    #              default orientation is 'S'
    #              for 'X', the filter is inactive and just returns the input.
    #   df1dx, df1dxi etc.:  design sensitivity (2D arrays)
    #INTERNAL SETTINGS
    P = 40
    ep = 1e-4 
    xi_0 = 0.5 # parameters for smooth max/min functions
    # INPUT CHECKS
    if baseplate=='X':
    # bypass option: filter does not modify the blueprint design
        xi = x
        varargout = args
        return xi, varargout
    baseplateUpper = baseplate.upper()
    orientation = "SWNE"
    nRot = orientation.find(baseplateUpper) 
    nSens = max(0, len(args))

    # Orientation
    x = np.rot90(x, nRot, axes=(0,2))
    xi = np.zeros(np.shape(x))
    lstArgs = list(args)
    i = 0
    for arg in lstArgs:
        arg = np.rot90(arg, nRot,axes=(0,2))   
        lstArgs[i] = arg
        i = i+1
    args = tuple(lstArgs)
    nelz, nely, nelx = np.shape(x)

    #AM Filter
    Ns = 5
    Q = P + np.log(Ns)/np.log(xi_0)             
    SHIFT = 100*np.finfo(float).tiny **(1/P)
#     SHIFT = 100*np.finfo(float).eps **(1/P)
    BACKSHIFT = 0.95*Ns**(1/Q)*SHIFT**(P/Q)     
    Xi = np.zeros((nelz,nely,nelx))
    keep = np.zeros((nelz, nely, nelx))
    sq = np.zeros((nelz, nely, nelx)) 

    # Baseline: identity
    xi[:,nely-1,:] = x[:,nely-1,:]
    for i in range(nely - 2, -1, -1): #nely-2,nely-3,  0
###
#     xi[:,0,:] = x[:,0,:]
#     for i in range(1, nely+1, 1): #1,2,  nely
###
        # compute maxima of current base row
        cbr = np.zeros((nelz + 2, 1, nelx + 2))+ SHIFT
        cbr[1:nelz+1,0,1:nelx+1] = xi[:,i+1,:] + SHIFT
                        
###
        keep[:, i,:] = (cbr[1:(nelz+1), 0, 0:nelx]**P + cbr[1:(nelz+1), 0, 1:(nelx+1)]**P + cbr[1:(nelz+1), 0, 2:(nelx+2)]**P +
                        cbr[0:nelz, 0, 1:(nelx+1)]**P + cbr[2:(nelz+2), 0, 1:(nelx+1)]**P)
        Xi[:, i,:] = keep[:, i,:]**(1 / Q) - BACKSHIFT
        sq[:, i,:] = np.sqrt((x[:, i,:] - Xi[:, i,:])** 2 + ep)
        xi[:,i,:] = 0.5*((x[:,i,:]+Xi[:,i,:])-sq[:,i,:]+np.sqrt(ep))
            
    # SENSITIVITIES
    if nSens != 0:
        dfxi = deepcopy(list(args))
        dfx = deepcopy(list(args))
        varLambda = np.zeros((nelz, nSens, nelx))
        
        # from top to base layer
        for i in range(nely - 1): # 0, ,nely-2
            # smin sensitivity terms
            dsmindx = 0.5 * (1 - (x[:, i,:] - Xi[:, i,:]) / sq[:, i,:])
            dsmindXi = 1 - dsmindx
            
            # smax sensitivity terms
            cbr = np.zeros((nelz + 2, 1, nelx + 2)) + SHIFT
            #+ SHIFT 看看能否避免nan
            cbr[1:nelz + 1, 0, 1:nelx + 1] = xi[:, i + 1,:] + SHIFT            

     
            dmx = np.zeros((nelz, Ns, nelx))
            for j in range(5):
                if j <= 2:
                    dmx[:, j,:] = (P / Q) * keep[:, i,:]**(1 / Q - 1) * cbr[1:(nelz+1), 0, j:(nelx+j)]**(P - 1)
                elif j == 3:
                    dmx[:, j,:] = (P / Q) * keep[:, i,:]**(1 / Q - 1) * cbr[0:nelz, 0, 1:(nelx+1)]**(P - 1)
                elif j == 4:
                    dmx[:, j,:] = (P / Q) * keep[:, i,:]**(1 / Q - 1) * cbr[2:(nelz+2), 0, 1:(nelx+1)]**(P - 1)

### padding for dmx
#             dmx_padding=mp.pad(dmx,(1,1),'constant')
            dmx_padding=np.zeros((nelz+2, Ns, nelx+2))
            dmx_padding[1:(nelz+1),:,1:(nelx+1)]=dmx[:,:,:]

###
        
            for k in range(nSens):
                dfx[k][:, i,:] = dsmindx * (dfxi[k][:, i,:] + varLambda[:, k,:])
                preLambda = np.zeros((nelz + 2, 1, nelx + 2))
                preLambda[1:(nelz + 1), 0, 1:(nelx + 1)] = (dfxi[k][:, i,:] + varLambda[:, k,:]) * dsmindXi
                for nz in range(nelz):
                    for nx in range(nelx):
                        varLambda[nz, k, nx] = np.sum(
                            
                                      np.array([  preLambda[nz + 1, 0, nx ]*dmx_padding[nz+ 1, 2, nx],
                                               preLambda[nz + 1, 0, nx + 1]*dmx_padding[nz + 1, 1, nx + 1],
                                               preLambda[nz + 1, 0, nx + 2]*dmx_padding[nz + 1, 0, nx + 2],
                                               preLambda[nz , 0, nx + 1]*dmx_padding[nz, 4, nx+ 1],
                                               preLambda[nz + 2, 0, nx + 1]*dmx_padding[nz+2 , 3, nx + 1]    ])
                                             )
                                      
        i = nely - 1
        for k in range(nSens):
                dfx[k][:, i,:] = dfxi[k][:, i,:] + varLambda[:, k,:]

    # ORIENTATION
    xi = np.rot90(xi,-nRot, axes=(0,2))
    varargout = ()
    for s in range(nSens):
        varargout = varargout + (np.rot90(dfx[s],-nRot,axes=(0,2)) ,)

    return xi, varargout

File: AMFilter3D.m
function [ xi, varargout ] = AMFilter3D( x, baseplate, varargin )
%AMFILTER Applies a virtual additive manufacturing process to a 
%         2D blueprint design input.
%   Possible uses:
%   xi = AMfilter(x)              design transformation, default orientation
%   xi = AMfilter(x, baseplate)   idem, with baseplate orientation specified
%   [xi, df1dx, df2dx,...] = AMfilter(x, baseplate, df1dxi, df2dxi, ...)
%       This includes also the transformation of design sensitivities
% where
%   x : blueprint design (2D array), 0 <= x(i,j) <= 1
%   xi: printed design (2D array)
%   baseplate: character indicating baseplate orientation: 'N','E','S','W'
%              default orientation is 'S'
%              for 'X', the filter is inactive and just returns the input.
%   df1dx, df1dxi etc.:  design sensitivity (2D arrays)

%INTERNAL SETTINGS
P = 40; ep = 1e-4; xi_0 = 0.5; % parameters for smooth max/min functions

%INPUT CHECKS
if nargin==1, baseplate='S'; end 
if baseplate=='X', 
    % bypass option: filter does not modify the blueprint design
    xi = x;
    varargout = varargin;
    return;
end 
nRot=find(upper(baseplate)=='SWNE')-1;
nSens=max(0,nargin-2); 
if nargout~=nSens+1, error('Input/output arguments mismatch.'); end

%ORIENTATION
x=rot90(x,nRot);
xi=zeros(size(x));
for s=1:nSens
    varargin{s}=rot90(varargin{s},nRot);    
end
[nely,nelx, nelz]=size(x); 

%AM FILTER =====================
Ns=3;
Q=P+log(Ns)/log(xi_0); 
SHIFT = 100*realmin^(1/P); % small shift to prevent division by 0
BACKSHIFT = 0.95*Ns^(1/Q)*SHIFT^(P/Q);
XiX=zeros(size(x)); keepX=zeros(size(x)); sqX=zeros(size(x));
XiY=zeros(size(x)); keepY=zeros(size(x)); sqY=zeros(size(x));
% baseline: identity
xi(:,:,nelz)=x(:,:,nelz); % copy base face as-is
xiY = zeros(size(x));
xiX = zeros(size(x));

for i=(nelz-1):-1:1
    for j = 1:nely
    % compute maxima of current base row
    cbr = [0, xi(j,:,i+1), 0] + SHIFT; % pad with zeros
    keepY(j,:,i) = (cbr(1:nelx).^P + cbr(2:(nelx+1)).^P + cbr(3:end).^P);
    XiY(j,:,i) = keepY(j,:,i).^(1/Q) - BACKSHIFT;
    sqY(j,:,i) = sqrt((x(j,:,i)-XiY(j,:,i)).^2 + ep);
    % set row above to supported value using smooth minimum:
    xiY(j,:,i) = 0.5*((x(j,:,i)+XiY(j,:,i)) - sqY(j,:,i) + sqrt(ep));
    end
    
    for j = 1:nelx
    % compute maxima of current base row
    cbr = [0; xi(:,j,i+1); 0] + SHIFT; % pad with zeros
    keepX(:,j,i) = (cbr(1:nely).^P + cbr(2:(nely+1)).^P + cbr(3:end).^P);
    XiX(:,j,i) = keepX(:,j,i).^(1/Q) - BACKSHIFT;
    sqX(:,j,i) = sqrt((x(:,j,i)-XiX(:,j,i)).^2 + ep);
    % set row above to supported value using smooth minimum:
    xiX(:,j,i) = 0.5*((x(:,j,i)+XiX(:,j,i)) - sqX(:,j,i) + sqrt(ep));
    end
end

%SENSITIVITIES
if nSens
    dfxiCol=varargin; dfxCol=varargin;
    dfxiRow=varargin; dfxRow=varargin;
    
    for j = 1:nelx
        lambdaRow=zeros(nSens,nely); 
        % from top to base layer:
        for i=1:nelz-1
            % smin sensitivity terms
            dsmindx  = .5*(1-(x(:,j,i)-XiX(:,j,i))./sqX(:,j,i));
            %dsmindXi = .5*(1+(x(i,:)-Xi(i,:))./sq(i,:)); 
            dsmindXi = 1-dsmindx; 
            % smax sensitivity terms
            cbr = [0; xi(:,j,i+1); 0] + SHIFT; % pad with zeros
            dmx = zeros(Ns,nelx);
            for s=1:Ns
                dmx(s,:) = (P/Q)*keepX(:,j,i).^(1/Q-1).*cbr((1:nely)+(s-1)).^(P-1);
            end        
            % rearrange data for quick multiplication:
            qj=repmat([-1 0 1]',nely,1);
            qi=repmat(1:nely,3,1); qi=qi(:);
            qj=qj+qi; qs=dmx(:);
            dsmaxdxi=sparse(qi(2:end-1),qj(2:end-1),qs(2:end-1)); 
            for k=1:nSens
                dfxRow{k}(:,j,i) = dsmindx.*(dfxiRow{k}(:,j,i)+lambdaRow(k,:)');
                lambdaRow(k,:)= ((dfxiRow{k}(:,j,i)+lambdaRow(k,:)').*dsmindXi)'*dsmaxdxi;
            end
        end
        % base layer:
        i=nelz;
        for k=1:nSens
            dfxRow{k}(:,j,i) = dfxiCol{k}(:,j,i)+lambdaRow(k,:)';
        end
    end
    
    for j = 1:nely
        lambdaCol=zeros(nSens,nelx); 
        % from top to base layer:
        for i=1:nelz-1
            % smin sensitivity terms
            dsmindx  = .5*(1-(x(j,:,i)-XiX(j,:,i))./sqX(j,:,i));
            %dsmindXi = .5*(1+(x(i,:)-Xi(i,:))./sq(i,:)); 
            dsmindXi = 1-dsmindx; 
            % smax sensitivity terms
            cbr = [0, xi(j,:,i+1), 0] + SHIFT; % pad with zeros
            dmx = zeros(Ns,nely);
            for s=1:Ns
                dmx(s,:) = (P/Q)*keepY(j,:,i).^(1/Q-1).*cbr((1:nelx)+(s-1)).^(P-1);
            end        
            % rearrange data for quick multiplication:
            qj=repmat([-1 0 1]',nelx,1);
            qi=repmat(1:nelx,3,1); qi=qi(:);
            qj=qj+qi; qs=dmx(:);
            dsmaxdxi=sparse(qi(2:end-1),qj(2:end-1),qs(2:end-1)); 
            for k=1:nSens
                dfxCol{k}(j,:,i) = dsmindx.*(dfxiCol{k}(j,:,i)+lambdaCol(k,:));
                lambdaCol(k,:)= ((dfxiCol{k}(j,:,i)+lambdaCol(k,:)).*dsmindXi)*dsmaxdxi;
            end
        end
        % base layer:
        i=nelz;
        for k=1:nSens
            dfxCol{k}(:,j,i) = dfxCol{k}(:,j,i)+lambdaCol(k,:)';
        end
    end
    
    for k = 1:nSens
        dfx{k} = 0.5*(dfxCol{k}+dfxRow{k});
    end
end

%ORIENTATION
xi=rot90(xi,-nRot);
for s=1:nSens
    varargout{s}=rot90(dfx{s},-nRot);    
end

end
%
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
% This Matlab code was written by Matthijs Langelaar,                      %
% Department of Precision and Microsystems Engineering,                    %
% Delft University of Technology, Delft, the Netherlands.                  %
% Please sent your comments to: m.langelaar@tudelft.nl                     %
%                                                                          %
% The code is intended for educational purposes and theoretical details    %
% are discussed in the paper "An additive manufacturing filter for         %
% topology optimization of print-ready designs", M. Langelaar (2016),      %
% Struct Multidisc Optim, DOI: 10.1007/s00158-016-1522-2.                  %
%                                                                          %
% This code is intended for integration in the 88-line topology            %
% optimization code discussed in the paper                                 %
% "Efficient topology optimization in MATLAB using 88 lines of code,       %
% E. Andreassen, A. Clausen, M. Schevenels, B. S. Lazarov and O. Sigmund,  % 
% Struct Multidisc Optim, 2010, Vol 21, pp. 120--127.                      %
%                                                                          %
% Disclaimer:                                                              %
% The author reserves all rights but does not guarantee that the code is   %
% free from errors. Furthermore, the author shall not be liable in any     %
% event caused by the use of the program.                                  %
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%